### DISCUSSION TITLE
Improved git completions

### DISCUSSION THREAD

**Discussion by drbrain** [nushell/nushell#6671] [Show and tell]
I've been working on completions for `git` (and other commands) in my [nushell configs](https://github.com/drbrain/nushell-config).

Lately I [added a parser](https://github.com/drbrain/nushell-config/blob/main/git_wrapper.nu) for [`git status --porcelain=2`](https://git-scm.com/docs/git-status#_porcelain_format_version_2) which is my first time actually writing a program in nushell, and maybe there ares some pointers you could give to make it better?

* Is there a `match` or `case` that would make the `if`/`else if`/… more concise?
* Is there a better way of building up the records in `parse_line`?

The parser lets me `git add <TAB>` and see only the files that are eligible for addition to the repository.

Here's some example output from the wrapper:

```
❯ git_wrapper status
╭───┬────────────────────┬───────────┬────────────┬──────────┬───────────┬────────────┬───────────────┬───────────────────────────┬──────────────────────────╮
│ # │        name        │  status   │   staged   │ unstaged │ mode_head │ mode_index │ mode_worktree │         name_head         │        name_index        │
├───┼────────────────────┼───────────┼────────────┼──────────┼───────────┼────────────┼───────────────┼───────────────────────────┼──────────────────────────┤
│ 0 │ completions/git.nu │ changed   │ unmodified │ modified │ 100644    │ 100644     │ 100644        │ 7e493b61ddee0d935e20f1620 │ 7e493b61ddee0d935e20f162 │
│   │                    │           │            │          │           │            │               │ c73521f10b6ac77           │ 0c73521f10b6ac77         │
│ 1 │ README.md          │ untracked │ ❎         │ ❎       │ ❎        │ ❎         │ ❎            │ ❎                        │ ❎                       │
╰───┴────────────────────┴───────────┴────────────┴──────────┴───────────┴────────────┴───────────────┴───────────────────────────┴──────────────────────────╯
```

`git add <TAB>`:

```
❯ | git add
completions/git.nu  README.md
```

↑ maybe my prompt is making that extra `|` appear?

And `git status --porcelain=2`

```
❯ ^git status --porcelain=2
1 .M N... 100644 100644 100644 7e493b61ddee0d935e20f1620c73521f10b6ac77 7e493b61ddee0d935e20f1620c73521f10b6ac77 completions/git.nu
? README.md
```

Someday I'll add some args support to allow including ignored files, or removing untracked files so `git checkout` can complete correctly.

**Comment by fdncred** [comment]
Here's a `match` custom command that someone wrote. I've also used it in the `register-plugins.nu` in the root of the repo.

```
# match
def match [input, matchers: record] {
    echo $matchers | get $input | do $in
}

# match example
# > let x = "3"

# match $x {
#   1: { echo "you chose one" },
#   2: { echo "you chose two" },
#   3: { echo "you chose three" }
# }
```

**Reply by drbrain** [reply]
That `match` is nice, I incorporated it and it cleans some things up nicely.

I can't figure out a way to pass something like `--ignored` from `completions_git git status` to `git_wrapper status` that doesn't involve a lot of argument wrangling. I see I need to use `run-external --redirect-stdout` but I'm not fond of having five lines of `if` block for every argument so maybe there's something I'm missing?

Here are what I have for [`git status`](https://github.com/drbrain/nushell-config/blob/df35c9d8bd10084f4780cc44dbb92eb6d4873e87/completions/git.nu#L795-L797) and [`git_wrapper status`](https://github.com/drbrain/nushell-config/blob/df35c9d8bd10084f4780cc44dbb92eb6d4873e87/git_wrapper.nu#L101-L113) extracted so you don't have to go hunting:

```nu
module complete_git {
  export def "git status" [--ignored] {
    git_wrapper status $ignored | select name status staged unstated --ignore-errors
  }
}

module git_wrapper {
  export def status [ignored: bool] {
    let args = ["status", "--porcelain=2"]
    let args = if $ignored {
      ($args | append "--ignored")
    } else {
      $args
    }

    ( run-external --redirect-stdout "git" $args
    | lines
    | each { |line| parse_line $line }
    )
  }
}
```
